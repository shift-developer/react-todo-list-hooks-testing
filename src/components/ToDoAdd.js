import React from 'react'
import { useForm } from './hooks/useForm';

export const ToDoAdd = ({handleAddToDo}) => {

    const [ { description }, handleInputChange, reset] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1) {
            return;
        }

        handleAddToDo({id: new Date().getTime(), description, done: false});
        reset();
    }
    
    return (
        <>
            <h4>Agregar ToDos</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input 
                    type="text"
                    name="description"
                    value={description}
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    onChange={handleInputChange}
                />
                <button className="btn btn-outline-primary mt-1 btn-block">
                    Agregar
                </button>
            </form> 
        </>
    )
}
