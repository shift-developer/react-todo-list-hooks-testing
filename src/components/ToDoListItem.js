import React from 'react';
import './styles/styles.css';

export const ToDoListItem = ({toDo, index, handleDelete, handleToggle}) => {
    return (
        <li
            key={toDo.id}
            className="list-group-item"
        >
            <p className={toDo.done ? "complete" : ""} onClick={ () => handleToggle(toDo.id)} >{index + 1}. {toDo.desc}</p>
            <button className="btn btn-danger" onClick={() => handleDelete(toDo.id)}>
                Borrar
            </button>
        </li>
    )
}
