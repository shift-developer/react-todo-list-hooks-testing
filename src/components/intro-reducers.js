const initialState = [
    {
        id: 1,
        todo: 'Comprar Carne',
        done: false
    }
];

const toDoReducer = (state = initialState, action) => {

    if(action?.type === 'agregar') {
        return [...state , action.payload ];
    }
    
    return state;
}

let toDos = toDoReducer();

const newToDo = 
{
        id: 2,
        todo: 'Comprar Leche',
        done: false
};

const agregarToDoAction = {
    type: 'agregar',
    payload: newToDo
};

toDos = toDoReducer( toDos, agregarToDoAction );

console.log(toDos)