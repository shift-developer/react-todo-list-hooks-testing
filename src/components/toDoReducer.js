
export const toDoReducer = (state = [], action) => {

    switch (action.type) {

        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter( toDo => toDo.id !== action.payload);
        case 'toggle':
            return state.map( toDo => (toDo.id === action.payload) ? {...toDo, done: !toDo.done} : toDo)
        default:
            return state;
    }
}