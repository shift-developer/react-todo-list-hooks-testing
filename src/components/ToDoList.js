import React from 'react';
import { ToDoListItem } from './ToDoListItem';

export const ToDoList = ({toDos, handleDelete, handleToggle}) => {
    return (
        <>
            {toDos.map( (toDo, index) => (
                <ToDoListItem 
                    key={toDo.id}
                    handleDelete={handleDelete} 
                    handleToggle={handleToggle} 
                    toDo={toDo} 
                    index={index}
                />
            ))}
        </>
    )
}
