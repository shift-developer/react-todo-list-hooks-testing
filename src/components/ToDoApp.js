import React, { useReducer } from 'react'
import { toDoReducer } from './toDoReducer';
import { useForm } from './hooks/useForm';
import './styles/styles.css';

const initialState = [
    {
        id: new Date().getTime(),
        description: "Aprender React",
        done: false
    }
]

export const ToDoApp = () => {

    const [ toDos, dispatch ] = useReducer(toDoReducer, initialState);

    const [ { description }, handleInputChange, reset] = useForm({
        description: ""
    })

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1) {
            return;
        }
        const newToDo = {id: new Date().getTime(), description, done: false};

        const action = {
            type: "add",
            payload: newToDo
        };

        dispatch(action);
        reset();
    }

    return (
        <div>
            <h1>ToDo App ( {toDos.length} )</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <ul>
                        {toDos.map( (item, i) => (
                            <li
                                key={item.id}
                                className="list-group-item"
                            >
                                <p className="text-center">{i + 1}. {item.description}</p>
                                <button className="btn btn-danger">
                                    Borrar
                                </button>
                            </li>
                        ))}
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar ToDos</h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input 
                            type="text"
                            name="description"
                            value={description}
                            className="form-control"
                            placeholder="Aprender..."
                            autoComplete="off"
                            onChange={handleInputChange}
                        />
                        <button className="btn btn-outline-primary mt-1 btn-block">
                            Agregar
                        </button>
                    </form> 
                </div>
            </div>

            
        </div>
    )
}
