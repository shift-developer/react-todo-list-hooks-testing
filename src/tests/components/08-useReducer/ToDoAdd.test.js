import React from 'react'
import {shallow} from 'enzyme'
import { ToDoAdd } from '../../../components/ToDoAdd'


describe('Pruebas sobre <ToDoAdd/>', () => {

    const handleAddToDo = jest.fn()

    const wrapper = shallow(
        <ToDoAdd
            handleAddToDo={handleAddToDo}
        />
    )
    
    test('debe de mostrarse correctamente ', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('NO debe de llamar handleAddToDo', () => {
        
        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){}});

        expect(handleAddToDo).toHaveBeenCalledTimes(0);
    })

    test('debe de llamar la funcion handleAddToDo', () => {
        const value = 'Aprender react';

        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        })

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){}});

        expect(handleAddToDo).toHaveBeenCalledTimes(1);
        expect(handleAddToDo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddToDo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });

        expect(wrapper.find('input').prop('value')).toBe('')

    })
    
})
