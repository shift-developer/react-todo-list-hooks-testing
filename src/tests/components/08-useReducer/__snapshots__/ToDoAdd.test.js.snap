// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <ToDoAdd/> debe de mostrarse correctamente  1`] = `
<Fragment>
  <h4>
    Agregar ToDos
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
